### PR TITLE
New Element: 2's Complement

### LINKED ISSUE #1373: Complementer Curcuit Element
**Is your feature request related to a problem? Please describe.**
While developing circuits, a lot of algorithms require 1's complement and 2's complements. It is really troublesome to create it manually, especially when using non-unity bid width.

**Describe the solution you'd like**
A clean,nice little element which returns the 2's complement of n-bit width line

**Describe alternatives you've considered**
A new element should be created

**Additional context**
Currently, we can use the not gate to calculate 1's complement, but to calculate 2's complement, we need to manually add one to it. Considering that this is used in a number of circuits, there shall be a dedicated element for it, just like an adder.

**Are you working on this? (Yes/No)**
Yes

### PR DESCRIPTION
**Pull request by YashKumarVerma**
Fixes #1373 

#### Describe the changes you have made in this PR -
- Added a new circuit element : **TwoComplement**
- Finds 2's complement of data stream (n-bit)

### Screenshots of the changes (If any) -
- Circuit Element in Misc Tab
image
- Element
image
- Attached elements
image

---
## Demonstration
circuit

### CODE REVIEW

File: public/js/modules.js [hunk 1]
```diff
@@ -1390,6 +1390,58 @@ Adder.prototype.resolve = function () {
...
+
+    var xx = this.x;
+    var yy = this.y;
+    ctx.beginPath();
+    ctx.fillStyle = "black";
+    fillText(ctx, "2'", xx, yy, 10);
+    if ((this.hover && !simulationArea.shiftDown) || simulationArea.lastSelected == this || simulationArea.multipleObjectSelections.contains(this)) ctx.fillStyle = "rgba(255, 255, 32,0.8)";
+    ctx.fill();
+    ctx.beginPath();
+    drawCircle2(ctx, 5,0, 15, xx, yy, this.direction);
```
[the remark this answers is not in the document]
Comment (YashKumarVerma) [reply public/js/modules.js]: outdated file, now resolved.

File: public/js/modules.js [hunk 2]
```diff
@@ -1390,6 +1390,58 @@ Adder.prototype.resolve = function () {
...
+    return data;
+}
+TwoComplement.prototype.resolve = function () {
+    if (this.isResolvable() == false) {
+        return;
+    }
+    let output = ((~this.inp1.value >>> 0) << (32 - this.bitWidth)) >>> (32 - this.bitWidth);
+    output = output + 1;
+    this.output1.value = ((output) << (32 - this.bitWidth)) >>> (32 - this.bitWidth);
+simulationArea.simulationQueue.add(this.output1);
```
[the remark this answers is not in the document]
Comment (YashKumarVerma) [reply public/js/modules.js]: outdated file, now resolved.

File: public/js/modules.js [hunk 3]
```diff
@@ -1390,6 +1390,58 @@ Adder.prototype.resolve = function () {
...
+        },
+    }
+    return data;
+}
+TwoComplement.prototype.resolve = function () {
+    if (this.isResolvable() == false) {
+        return;
+    }
+    let output = ((~this.inp1.value >>> 0) << (32 - this.bitWidth)) >>> (32 - this.bitWidth);
+    output = output + 1;
```
[the remark this answers is not in the document]
Comment (YashKumarVerma) [reply public/js/modules.js]: outdated file, now resolved.

File: public/js/modules.js [hunk 4]
```diff
@@ -1390,6 +1390,58 @@ Adder.prototype.resolve = function () {
     simulationArea.simulationQueue.add(this.sum);
 }
 
+function TwoComplement(x, y, scope = globalScope, dir = "RIGHT", bitWidth = 1) {
```
[the remark this answers is not in the document]
Comment (YashKumarVerma) [reply public/js/modules.js:1393]: following standards as done for other elements. 5 arguments are required, therefore wontfix.

**Comment by mahmodHammad** [comment]
Good job!👏

**Comment by yash15601** [comment]
Good work yash😃

**Comment by Shivansh2407** [comment]
Nice @YashKumarVerma 👍 . Wanted to know that I am trying to reproduce the circuit in the Attached Elements screenshot and I am getting a bitwidth error. Can you please check once.

**Comment by YashKumarVerma** [comment]
> Nice @YashKumarVerma 👍 . Wanted to know that I am trying to reproduce the circuit in the Attached Elements screenshot and I am getting a bitwidth error. Can you please check once.

check if you have changed the bit-width of all the elements

**Comment by Shivansh2407** [comment]
Yes I have changed the bitwidth of all the elements but as soon as I change the bitwidth of Two's Complement, it does not sets the bitwidth and looks something like the screenshot below. Please check once .
image
It reverts the bitwidth of two's complement to null / ' ' .

**Comment by YashKumarVerma** [comment]
circuit

This should clear it up I guess.

**Comment by Ananthavijay** [comment]
@YashKumarVerma I think the color property is missing for the 2's Complement

File: public/js/modules.js [hunk 5]
```diff
@@ -1390,6 +1390,58 @@ Adder.prototype.resolve = function () {
...
+            inp1: findNode(this.inp1)
+        },
+    }
+    return data;
+}
+TwoComplement.prototype.resolve = function () {
+    if (this.isResolvable() == false) {
+        return;
+    }
+    let output = ((~this.inp1.value >>> 0) << (32 - this.bitWidth)) >>> (32 - this.bitWidth);
```
Comment (satu0king) [inline public/js/modules.js:1421]: Isn't this just -this.inp1.value ?

**Comment by satu0king** [comment]
Works great

**Comment by satu0king** [comment]
The look and SVG can definitely be improved in a later PR!

**Pull request merged**